test/Chainweb/Test/Pact/Checkpointer.hs

```diff
@@ -310,3 +326,409 @@ testCheckpointer loggers CheckpointEnv{..} dbState00 = do
+
+
+  discardNew _cpeCheckpointerNew >>= assertEitherSuccess "discard for block 02"
+
+  ------------------------------------------------------------------
+  -- s07 : validate block 03
+  ------------------------------------------------------------------
+
+  hash03 <- BlockHash <$> merkleLogHash "0000000000000000000000000000003a"
+
+  blockenv13 <- restoreNew _cpeCheckpointerNew bh03 (Just hash02) >>= assertEitherSuccess "restore for validate block 03"
+
+  withBlockEnv blockenv13 $ \benv -> do
+
+    -- insert here would fail if new block 03 had not been discarded
+    void $ runExec chainwebpactdb benv Nothing "(m2.insertTbl 'b 2)"
+
+    runExec chainwebpactdb benv Nothing "(m2.readTbl)"
+      >>= \EvalResult{..} -> Right _erOutput @?= traverse toPactValue [tIntList [1,2]]
+
+
+    saveNew _cpeCheckpointerNew bh03 hash03 >>= assertEitherSuccess "save block 01"
+
+------------------------------------------------------------------
+  -- s08: FORK! block 02, new hash
+  -- recreate m2 module, exercise RefStore checkpoint
+  -- exec next part of pact
+  ------------------------------------------------------------------
+
+  hash02Fork <- BlockHash <$> merkleLogHash "0000000000000000000000000000002b"
+
+  blockenv02Fork <- restoreNew _cpeCheckpointerNew bh02 (Just hash01) >>= assertEitherSuccess "restore for validate block 02"
+
+  withBlockEnv blockenv02Fork $ \benv -> do
+
+    void $ runExec chainwebpactdb benv (Just $ ksData "2") $ defModule "2"
+
+    runExec chainwebpactdb benv Nothing "(m2.readTbl)" >>= \EvalResult{..} -> Right _erOutput @?= traverse toPactValue [tIntList [1]]
+
+    -- this would fail if not a fork
+    runCont chainwebpactdb benv pactId 1 >>= ((Just 1 @=?) . pactCheckStep)
+
+    saveNew _cpeCheckpointerNew bh02 hash02Fork >>= assertEitherSuccess "save block 02"
+
+toTerm' :: ToTerm a => a -> Term Name
+toTerm' = toTerm
+
+{- You should probably think about the orphan, their name is
+ ExecutionMode, that you are leaving behind. They deserve a home! -}
+_testRegress :: Assertion
+_testRegress =
+  regressChainwebPactDb >>= fmap (toTup . _benvBlockState) . readMVar >>=
+  assertEquals "The final block state is" finalBlockState
+  where
+    finalBlockState = (2, BlockVersion 0 0, M.empty)
+    toTup (BlockState txid _ blockVersion txRecord _) =
+      (txid, blockVersion, txRecord)
+
+regressChainwebPactDb :: IO (MVar (BlockEnv SQLiteEnv))
+regressChainwebPactDb = do
+ withTempSQLiteConnection []  $ \sqlenv -> do
+        let initBlockState = BlockState 0 Nothing (BlockVersion 0 0) M.empty (newLogger loggers "BlockEnvironment")
+            loggers = pactTestLogger False
+        runRegression chainwebpactdb
+          (BlockEnv sqlenv initBlockState)
+          (\v -> runBlockEnv v initSchema)
+
+begin :: PactDb e -> Method e (Maybe TxId)
+begin pactdb = _beginTx pactdb Transactional
+
+commit :: PactDb e -> Method e [TxLog Value]
+commit pactdb = _commitTx pactdb
+
+runRegression ::
```

Comment:
comment that this should be migrated back to Pact

Reply:
I actually did this before you wrote the comment. I should have uploaded that commit before you started this review.